crates/openid4vci/src/credential_issuer/mod.rs

```diff
@@ -372,6 +363,36 @@ pub struct CNonce {
 pub struct EvaluateCredentialRequestOptions {
     /// Additional nonce options for validation
     pub c_nonce: Option<CNonceOptions>,
+
+    /// Id of the client that will be checked whether it is equal to the `iss` field inside the
+    /// `JWT` proof
+    ///
+    /// For the Pre-Authorized Code Grant Type, authentication of the client is OPTIONAL, as
+    /// described in Section 3.2.1 of OAuth 2.0 [RFC6749](https://www.rfc-editor.org/info/rfc6749)
+    /// and consequently, the "client_id" is only needed when a form of Client Authentication that
+    /// relies on the parameter is used.
+    ///
+    /// We deal with it being `OPTIONAL` by using the following algorithm:
+    ///
+    /// - If pre-authorized flow is used:
+    ///     - If `client_id` is supplied:
+    ///         - validate the `client_id` with the `iss` field in the `JWT`
+    ///     - if `client_id` is not supplied:
+    ///         - Do not validate even if the `iss` field is supplied within the `JWT`
+    /// - if authorized code flow is used:
+    ///     - validate the `client_id` with the `iss` field inside the `JWT`
+    pub client_id: Option<String>,
+
+    /// Whether the flow in which this function is called is the pre-authorized code flow. This is
+    /// required to see whether the `client_id` should be validated against the `iss` field inside
+    /// the `JWT`.
+    ///
+    /// if no options are supplied, we assume the stricted and validate the `iss` inside the `JWT`
+    /// with the `client_id`.
+    ///
+    /// TODO: Is there a way to omit this and infer it from the rest of the params? I.e. credential
+    /// offer vs authorization server metadata is supplied.
+    pub is_pre_authorized_code_flow: bool,
```

Comment:
We can determine this based on the grants from the credential offer and then what is requested in the authorization endpoint / token response.

This is how Sphereon handles it: https://github.com/Sphereon-Opensource/OID4VCI/blob/develop/packages/issuer/lib/VcIssuer.ts#L151-L174 (might be nice to look at their implementation in general and see what kinds of checks they do)

It seems they take the grants from the credential offer (https://github.com/Sphereon-Opensource/OID4VCI/blob/develop/packages/issuer/lib/VcIssuer.ts#L111), but I'm not sure what that means if both grants were used in the credential offer (you don't really know which grant was used). Maybe you can make sense of the implementation :)


Or -- we just support pre-auth, so we know what grant type is used (and we can keep it simple for now)